File: backend/F1Fantasy/Services/PaginationStateTracker.cs
using System.Collections.Concurrent;

namespace F1Fantasy.Services;

/// <summary>
/// Tracks pagination state for API endpoints to resume from last successful offset
/// when requests fail due to rate limiting or other errors
/// </summary>
public class PaginationStateTracker
{
    private readonly ConcurrentDictionary<string, PaginationState> _states = new();

    public class PaginationState
    {
        public int LastSuccessfulOffset { get; set; } = 0;
        public int Total { get; set; } = 0;
        public bool IsComplete { get; set; } = false;
        public DateTime LastUpdate { get; set; } = DateTime.UtcNow;
    }

    /// <summary>
    /// Get the current pagination state for an endpoint
    /// </summary>
    public PaginationState GetState(string endpoint)
    {
        return _states.GetOrAdd(endpoint, _ => new PaginationState());
    }

    /// <summary>
    /// Update the pagination state after a successful fetch
    /// </summary>
    public void UpdateState(string endpoint, int offset, int total)
    {
        var state = _states.GetOrAdd(endpoint, _ => new PaginationState());
        state.LastSuccessfulOffset = offset;
        state.Total = total;
        state.LastUpdate = DateTime.UtcNow;
        
        // Mark as complete if we've reached or exceeded the total
        const int limit = 30; // API page size
        state.IsComplete = (offset + limit) >= total;
    }

    /// <summary>
    /// Mark pagination as complete for an endpoint
    /// </summary>
    public void MarkComplete(string endpoint)
    {
        var state = _states.GetOrAdd(endpoint, _ => new PaginationState());
        state.IsComplete = true;
        state.LastUpdate = DateTime.UtcNow;
    }

    /// <summary>
    /// Reset pagination state for an endpoint (force refresh)
    /// </summary>
    public void Reset(string endpoint)
    {
        _states.TryRemove(endpoint, out _);
    }

    /// <summary>
    /// Check if we should attempt to fetch more data
    /// Returns true if incomplete or data is stale (older than 1 hour)
    /// </summary>
    public bool ShouldFetch(string endpoint)
    {
        if (!_states.TryGetValue(endpoint, out var state))
        {
            return true; // No state, should fetch
        }

        if (!state.IsComplete)
        {
            return true; // Incomplete, continue fetching
        }

        // If complete but data is older than 1 hour, refresh
        var dataAge = DateTime.UtcNow - state.LastUpdate;
        return dataAge.TotalHours > 1;
    }

    /// <summary>
    /// Get the next offset to fetch from
    /// </summary>
    public int GetNextOffset(string endpoint)
    {
        if (!_states.TryGetValue(endpoint, out var state))
        {
            return 0; // Start from beginning
        }

        if (state.IsComplete)
        {
            return 0; // Restart if complete and doing a refresh
        }

        // Continue from after the last successful offset
        const int limit = 30;
        return state.LastSuccessfulOffset + limit;
    }
}

File: backendJava/f1fantasy/src/main/java/no/f1fantasy/service/PaginationStateTracker.java
package no.f1fantasy.service;

import org.springframework.stereotype.Component;

import java.time.OffsetDateTime;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;

@Component
public class PaginationStateTracker {

    private final ConcurrentMap<String, PaginationState> states = new ConcurrentHashMap<>();

    public PaginationState getState(String endpoint) {
        return states.computeIfAbsent(endpoint, key -> new PaginationState());
    }

    public void updateState(String endpoint, int offset, int total, int limit) {
        PaginationState state = states.computeIfAbsent(endpoint, key -> new PaginationState());
        state.setLastSuccessfulOffset(offset);
        state.setTotal(total);
        state.setLastUpdate(OffsetDateTime.now());
        state.setComplete(offset + limit >= total);
    }

    public void markComplete(String endpoint) {
        PaginationState state = states.computeIfAbsent(endpoint, key -> new PaginationState());
        state.setComplete(true);
        state.setLastUpdate(OffsetDateTime.now());
    }

    public void reset(String endpoint) {
        states.remove(endpoint);
    }

    public boolean shouldFetch(String endpoint) {
        PaginationState state = states.get(endpoint);
        if (state == null) {
            return true;
        }

        if (!state.isComplete()) {
            return true;
        }

        return state.getLastUpdate().isBefore(OffsetDateTime.now().minusHours(1));
    }

    public int getNextOffset(String endpoint, int limit) {
        PaginationState state = states.get(endpoint);
        if (state == null || state.isComplete()) {
            return 0;
        }
        return state.getLastSuccessfulOffset() + limit;
    }

    public static class PaginationState {
        private int lastSuccessfulOffset;
        private int total;
        private boolean complete;
        private OffsetDateTime lastUpdate = OffsetDateTime.now();

        public int getLastSuccessfulOffset() {
            return lastSuccessfulOffset;
        }

        public void setLastSuccessfulOffset(int lastSuccessfulOffset) {
            this.lastSuccessfulOffset = lastSuccessfulOffset;
        }

        public int getTotal() {
            return total;
        }

        public void setTotal(int total) {
            this.total = total;
        }

        public boolean isComplete() {
            return complete;
        }

        public void setComplete(boolean complete) {
            this.complete = complete;
        }

        public OffsetDateTime getLastUpdate() {
            return lastUpdate;
        }

        public void setLastUpdate(OffsetDateTime lastUpdate) {
            this.lastUpdate = lastUpdate;
        }
    }
}
